ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -528,15 +543,47 @@ RouterException getChunkException() {
      *                                    created as part of this poll operation.
      */
     void poll(RequestRegistrationCallback<GetOperation> requestRegistrationCallback) {
-      //First, check if any of the existing requests have timed out.
-      cleanupExpiredInFlightRequests();
-      checkAndMaybeComplete();
-      if (!isComplete()) {
-        fetchRequests(requestRegistrationCallback);
+      if (!maybeProcessDecryptCallbackAndCompleteOperation()) {
+        //First, check if any of the existing requests have timed out.
+        cleanupExpiredInFlightRequests();
+        checkAndMaybeComplete();
+        if (!isComplete()) {
+          fetchRequests(requestRegistrationCallback);
+        }
       }
     }
 
     /**
+     * Maybe process decrypt callback and complete the operation if applicable. This is a no-op for non-encrypted blobs
+     * @return {@code true} if the operation is complete, {@code false} otherwise
+     */
+    protected boolean maybeProcessDecryptCallbackAndCompleteOperation() {
+      if (decryptionRequired.get() && decryptCallbackResultInfo.resultAvailable) {
```

Comment:
the `DecryptionTracker` can be leveraged here I think. 
Also, it might even be possible to move the check to outside this method so you can avoid the `maybe` in the method name. It might be easier to follow also.

Reply:
I would like to keep it this way. Infact I am planning to rename this method as mayBeProcessCallbacksAndCompleteOperation. In future, we should extend the same method to do any compression result processing as well. So, would prefer to keep it this way and not flood poll() with more details.